### PR TITLE
Add sitemap support

### LINKED ISSUE #11429: Generate a sitemap for SEO
Gitea could generate a sitemap for SEO, so crawlers can search through public repositories more easily. It could be useful for some admins that their instances will be crawled and ranked by search engines. A good explanation for the format of `sitemap.xml` can be found on [Wikipedia](https://en.wikipedia.org/wiki/Sitemaps).

In my opinion this feature should do the following:
- The sitemap should be available at the path `/sitemap.xml`.
- The sitemap should at least list the main page of each public repository and group. The main page of the wiki's / all wiki's pages could be optional.
- `lastmod` should be the date of the last commit to the README file / the wiki page itself

### PR DESCRIPTION
**Pull request by danog**
Fixes #11429.
This is my first time working with the codebase, I realize my approach may have been a a bit non-uniform at times, I'm open to suggestions!

### CODE REVIEW

File: modules/context/context.go [hunk 1]
```diff
@@ -208,6 +209,23 @@ func (ctx *Context) HTML(status int, name base.TplName) {
...
+	idx := sitemap.NewSitemapIndex()
+	count := 1
+	for i := 0; i < total; i += pagination {
+		idx.Add(&sitemap.URL{
+			Loc: base + "-" + strconv.Itoa(count) + ".xml",
+		})
+		count++
+	}
+	ctx.Resp.WriteHeader(http.StatusOK)
+	ctx.Resp.Header().Set("Content-Type", "text/xml")
```
Comment (KN4CK3R) [inline modules/context/context.go]:
```suggestion
	ctx.Resp.Header().Set("Content-Type", "text/xml")
```
That's the wrong order. After `WriteHeader` you can't change the header. The first `Write` calls `WriteHeader` with `StatusOK`,

File: routers/web/explore/user.go [hunk 2]
```diff
@@ -73,6 +79,25 @@ func RenderUserSearch(ctx *context.Context, opts *user_model.SearchUserOptions,
...
+			ctx.RenderSitemapPagination(setting.AppURL+"explore/sitemap", int(count), setting.UI.User.RepoPagingNum)
+		} else {
+			m := sitemap.NewSitemapIndex()
+			for _, item := range users {
+				m.Add(&sitemap.URL{
+					Loc: item.HTMLURL() + ".xml",
+				})
+			}
+			ctx.Resp.WriteHeader(http.StatusOK)
+			ctx.Resp.Header().Set("Content-Type", "text/xml")
```
Comment (KN4CK3R) [inline routers/web/explore/user.go]:
```suggestion
			ctx.Resp.Header().Set("Content-Type", "text/xml")
```
same

File: routers/web/home.go [hunk 3]
```diff
@@ -59,6 +60,22 @@ func Home(ctx *context.Context) {
...
+func HomeSitemap(ctx *context.Context) {
+	idx := sitemap.NewSitemapIndex()
+	if !setting.Service.Explore.DisableUsersPage {
+		idx.Add(&sitemap.URL{
+			Loc: setting.AppURL + "explore/users/sitemap-0.xml",
+		})
+	}
+
+	ctx.Resp.WriteHeader(http.StatusOK)
+	ctx.Resp.Header().Set("Content-Type", "text/xml")
```
Comment (KN4CK3R) [inline routers/web/home.go]:
```suggestion
	ctx.Resp.Header().Set("Content-Type", "text/xml")
```
same

File: routers/web/user/profile.go [hunk 4]
```diff
@@ -353,7 +374,24 @@ func Profile(ctx *context.Context) {
...
+	} else if isSubSitemap {
+		m := sitemap.New()
+		for _, item := range repos {
+			m.Add(&sitemap.URL{
+				Loc:     item.HTMLURL(),
+				LastMod: item.UpdatedUnix.AsTimePtr(),
+			})
+		}
+		ctx.Resp.WriteHeader(http.StatusOK)
+		ctx.Resp.Header().Set("Content-Type", "text/xml")
```
Comment (KN4CK3R) [inline routers/web/user/profile.go]:
```suggestion
		ctx.Resp.Header().Set("Content-Type", "text/xml")
```
same

**Comment by danog** [comment]
Cleaned up a bit :)

**Comment by danog** [comment]
Marking this as draft, it seems Google doesn't support nested sitemap indexes, I'll have to rewrite this.

**Comment by danog** [comment]
All done, defaulting to the (spec limit) of 50k URLs per sitemap.

File: routers/web/home.go [hunk 5]
```diff
@@ -59,6 +66,55 @@ func Home(ctx *context.Context) {
...
+			ListOptions: db.ListOptions{PageSize: 1},
+			IsActive:    util.OptionalBoolTrue,
+			Visible:     []structs.VisibleType{structs.VisibleTypePublic},
+		})
+		if err != nil {
+			ctx.ServerError("SearchUsers", err)
+			return
+		}
+		idx := 1
+		for i := int64(0); i < count; i += 50000 {
```
Comment (6543) [inline routers/web/home.go]: can this be abused to DOS gitea?

Comment (danog) [reply routers/web/home.go]: I had the same doubt when setting this initially, but I was also afraid of running out of subsitemaps due to a low paging size.
Then I realized that with the default explore page size of 20, even 100k repos + 100k users would only take up 10k subsitemaps, well below the limit of 50k.
Reverted to the default explore page size.

File: go.mod [hunk 6]
```diff
@@ -101,6 +101,7 @@ require (
 	github.com/sergi/go-diff v1.2.0
 	github.com/shurcooL/httpfs v0.0.0-20190707220628-8d4bc4ba7749 // indirect
 	github.com/shurcooL/vfsgen v0.0.0-20200824052919-0d455de96546
+	github.com/snabb/sitemap v1.0.0 // indirect
```
Comment (Gusted) [inline go.mod]: I think we should not import such a small library(that hasn't any test), I think we rather should make our own in modules/sitemap. ~~This depedencies imports another dependencies which we won't use as well.~~

File: routers/web/home.go [hunk 7]
```diff
@@ -59,6 +66,57 @@ func Home(ctx *context.Context) {
 	ctx.HTML(http.StatusOK, tplHome)
 }
 
+// Home renders main sitemap
+func HomeSitemap(ctx *context.Context) {
+	m := sitemap.NewSitemapIndex()
+	if !setting.Service.Explore.DisableUsersPage {
+		_, cnt, err := user_model.SearchUsers(&user_model.SearchUserOptions{
+			Actor:       ctx.User,
```
Comment (Gusted) [inline routers/web/home.go]: Is it feasible to have a scenario whereby a authenticated user is going to request sitemap.xml? I'd really only expect search machines is going trough here.

File: routers/web/web.go [hunk 8]
```diff
@@ -230,6 +230,9 @@ func RegisterRoutes(m *web.Route) {
 	// Routers.
 	// for health check
 	m.Get("/", Home)
+	m.Group("/sitemap.xml", func() {
+		m.Get("", HomeSitemap)
+	}, ignExploreSignIn)
```
Comment (Gusted) [inline routers/web/web.go]:
```suggestion
	m.Get("/sitemap.xml", ignExploreSignIn, HomeSitemap)
```
I don't see the benefit of using `Group` here.

File: routers/web/home.go [hunk 9]
```diff
@@ -59,6 +66,57 @@ func Home(ctx *context.Context) {
...
+		idx := 1
+		for i := 0; i < count; i += setting.UI.ExplorePagingNum {
+			m.Add(&sitemap.URL{
+				Loc: setting.AppURL + "explore/users/sitemap-" + strconv.Itoa(idx) + ".xml",
+			})
+			idx++
+		}
+	}
+
+	_, cnt, err := models.SearchRepository(&models.SearchRepoOptions{
```
Comment (Gusted) [inline routers/web/home.go]: Ditto.

**Comment by danog** [comment]
All done!
Wrote a simplified sitemap library with additional checks as per the specs+tests.

File: go.mod [hunk 10]
```diff
@@ -75,6 +75,7 @@ require (
 	github.com/santhosh-tekuri/jsonschema/v5 v5.0.0
 	github.com/sergi/go-diff v1.2.0
 	github.com/shurcooL/vfsgen v0.0.0-20200824052919-0d455de96546
+	github.com/ssor/bom v0.0.0-20170718123548-6386211fdfcf // indirect
```
Comment (6543) [inline go.mod]: looks unrelated

**Comment by 6543** [comment]
please do not force-push or squash, it will make review easyer & in the end it always get squash-merged :+1:

File: routers/web/home.go [hunk 11]
```diff
@@ -59,6 +66,53 @@ func Home(ctx *context.Context) {
...
+			m.Add(sitemap.URL{URL: setting.AppURL + "explore/users/sitemap-" + strconv.Itoa(idx) + ".xml"})
+			idx++
+		}
+	}
+
+	_, cnt, err := models.SearchRepository(&models.SearchRepoOptions{
+		ListOptions: db.ListOptions{
+			PageSize: 1,
+		},
+		Actor:     nil,
```
Comment (6543) [inline routers/web/home.go]:
```suggestion
		Actor:     ctx.Doer,
```

File: routers/web/home.go [hunk 12]
```diff
@@ -59,6 +66,53 @@ func Home(ctx *context.Context) {
...
+		AllPublic: true,
+	})
+	if err != nil {
+		ctx.ServerError("SearchRepository", err)
+		return
+	}
+	count := int(cnt)
+	idx := 1
+	for i := 0; i < count; i += setting.UI.ExplorePagingNum {
```
Comment (6543) [inline routers/web/home.go]: the ExplorePagingNum is not used for a max page size but for max item per page

**Comment by Tha14** [comment]
Any updates on this? I would really like to have sitemaps available. Thanks

File: modules/sitemap/sitemap.go [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+		return 0, fmt.Errorf("The sitemap contains too many URLs: %d", len(s.URLs))
+	}
+	buf := bytes.NewBufferString(xml.Header)
+	if err := xml.NewEncoder(buf).Encode(s); err != nil {
+		return 0, err
+	}
+	if err := buf.WriteByte('\n'); err != nil {
+		return 0, err
+	}
+	if buf.Len() > 50*1024*1024 {
```
Comment (Gusted) [inline modules/sitemap/sitemap.go]: `50*1024*1024` this should be a configurable constant.

File: routers/web/home.go [hunk 14]
```diff
@@ -59,6 +66,53 @@ func Home(ctx *context.Context) {
 	ctx.HTML(http.StatusOK, tplHome)
 }
 
+// HomeSitemap renders the main sitemap
+func HomeSitemap(ctx *context.Context) {
+	m := sitemap.NewSitemapIndex()
+	if !setting.Service.Explore.DisableUsersPage {
+		_, cnt, err := user_model.SearchUsers(&user_model.SearchUserOptions{
+			Actor:       nil,
```
Comment (Gusted) [inline routers/web/home.go]:
```suggestion
```

File: routers/web/home.go (hunk 12 above)
Comment (Gusted) [reply routers/web/home.go]: I think a new option along with ^^ the magic constant should be created for this PR.

File: modules/setting/setting.go [hunk 15]
```diff
@@ -207,6 +207,7 @@ var (
 	// UI settings
 	UI = struct {
 		ExplorePagingNum      int
+		SitemapPagingNum      int
```
Comment (6543) [inline modules/setting/setting.go:210]: please add a note into https://github.com/go-gitea/gitea/blob/main/docs/content/doc/advanced/config-cheat-sheet.en-us.md

**Review by 6543 — APPROVED** [review]

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 16]
```diff
@@ -174,6 +174,7 @@ The following configuration set `Content-Type: application/vnd.android.package-a
 - `MEMBERS_PAGING_NUM`: **20**: Number of members that are shown in organization members.
 - `FEED_MAX_COMMIT_NUM`: **5**: Number of maximum commits shown in one activity feed.
 - `FEED_PAGING_NUM`: **20**: Number of items that are displayed in home feed.
+- `SITEMAP_PAGING_NUM`: **20**: Number of items that are displayed in a single subsitemap.
```
Comment (Gusted) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md:177]: `custom/conf/app.example.ini` also needs a mention :)

**Review by Gusted — APPROVED** [review]

**Comment by 6543** [comment]
:rocket:

**Comment by danog** [comment]
Ayy, thanks for merging 🎉 
Do you think this could be backported/released in the upcoming 1.17?

**Comment by Gusted** [comment]
We don't backport features.

**Comment by danog** [comment]
> We don't backport features.

Ah well, guess I'll wait for 1.18 :)

**Review by flame4ost — COMMENTED** [review]
good
